oauth/api/router.py

```diff
@@ -0,0 +1,167 @@
+from logging import getLogger
+from typing import List
+
+from fastapi import APIRouter, Body, Depends, Request, Security, status
+from fastapi.security import HTTPBasic
+from sqlalchemy.orm import Session
+
+import oauth.api.endpoints as endpoints
+from oauth.api.exceptions import (
+    AuthenticationException,
+    ClientNotFoundException,
+    InvalidScopeException,
+)
+from oauth.api.models import (
+    AccessToken,
+    ClientCreatedResponse,
+    OAuth2ClientCredentialsRequestForm,
+)
+from oauth.api.utils import verify_oauth_client
+from oauth.database.models import ClientDetail
+from oauth.scopes import (
+    CLIENT_CREATE,
+    CLIENT_DELETE,
+    CLIENT_READ,
+    CLIENT_UPDATE,
+    SCOPE_READ,
+    SCOPES,
+)
+
+logger = getLogger(__name__)
+router = APIRouter(prefix=endpoints.OAUTH_PREFIX, tags=["OAuth"])
+
+
+@router.post(endpoints.TOKEN, response_model=AccessToken)
+async def acquire_access_token(
+    request: Request,
+    form_data: OAuth2ClientCredentialsRequestForm = Depends(),
+    db: Session = Depends(),
+) -> AccessToken:
+    """
+    Given a set of credentials, returns an access token.
+    """
+
+    basic_credentials = await HTTPBasic(auto_error=False)(request)
+
+    if form_data.client_id and form_data.client_secret:
+        client_id = form_data.client_id
+        client_secret = form_data.client_secret
+    elif basic_credentials:
+        client_id = basic_credentials.username
+        client_secret = basic_credentials.password
+    else:
+        raise AuthenticationException()
+
+    client_detail = ClientDetail.get(db, client_id=client_id)
+
+    if client_detail is None:
+        raise AuthenticationException()
+
+    if not client_detail.credentials_valid(client_secret):
+        raise AuthenticationException()
+
+    logger.info("Creating access token for client with ID '%s'", client_id)
+    return AccessToken(access_token=client_detail.create_access_code_jwe())
+
+
+@router.post(
+    endpoints.CLIENT,
+    dependencies=[Security(verify_oauth_client, scopes=[CLIENT_CREATE])],
+    response_model=ClientCreatedResponse,
+    status_code=status.HTTP_201_CREATED,
+)
+def create_client(
+    *,
+    db: Session = Depends(),
+    scopes: List[str] = Body([]),
+) -> ClientCreatedResponse:
+    """
+    Creates a new client and returns the credentials.
+    """
+
+    for scope in scopes:
+        if not scope in SCOPES:
+            raise InvalidScopeException(scope)
```

Comment:
I feel like there's a more concise/Python-y way to do this...

Reply:
Definitely feels like a list comprehension or some use of `set`

Reply:
This will only return the first invalid scope, it would be nice to do something like:
```
invalid_scopes = [scope for scope in scopes if scope not in SCOPES]
if len(invalid_scopes) >0:
    raise InvalidScopeException(invalid_scopes)
```